src/metabase/db/metadata_queries.clj

```diff
@@ -93,32 +93,47 @@
 
 (def TableRowsSampleOptions
   "Schema for `table-rows-sample` options"
-  (s/maybe {(s/optional-key :truncation-size) s/Int}))
+  (s/maybe {(s/optional-key :truncation-size) s/Int
+            (s/optional-key :rff)             s/Any}))
 
-(s/defn table-rows-sample :- (s/maybe si/TableSample)
-  "Run a basic MBQL query to fetch a sample of rows belonging to a Table. Can pass in optional `:truncation-size` for
-  text fields which will be honored only if the driver supports expressions."
+(defn- table-rows-sample-query
+  "Returns the mbql query to query a table for sample rows"
+  [table fields {:keys [truncation-size] :as _opts}]
+  (let [driver             (-> table table/database driver.u/database->driver)
+        text-fields        (filter (comp #{:type/Text} :base_type) fields)
+        field->expressions (when (and truncation-size (driver/supports? driver :expressions))
+                             (into {} (for [field text-fields]
+                                        [field [(str (gensym "substring"))
+                                                [:substring [:field-id (u/get-id field)] 1 truncation-size]]])))]
+    {:database   (:db_id table)
+     :type       :query
+     :query      {:source-table (u/get-id table)
+                  :expressions  (into {} (vals field->expressions))
+                  :fields       (vec (for [field fields]
+                                       (if-let [[expression-name _] (get field->expressions field)]
+                                         [:expression expression-name]
+                                         [:field-id (u/get-id field)])))
+                  :limit        max-sample-rows}
+     :middleware {:format-rows?           false
+                  :skip-results-metadata? true}}))
+
+(s/defn table-rows-sample
+  "Run a basic MBQL query to fetch a sample of rows of FIELDS belonging to a TABLE.
+
+  Options: a map of
+  `:truncation-size`: [optional] size to truncate text fields if the driver supports expressions.
+  `:rff`: [optional] a reducing function to reduce over the result set in the the query-processor rather than
+  realizing the whole collection"
```

Comment:
`rff` (at least elsewhere in the QP code) isn't a reducing function, it's a function that (given initial results metadata) returns a reducing function. If it was just a reducing function it would be called `rf` (conventional Clojure variable name for a *reducing function*). `rff` instead is short for *reducing function function*.

I don't know if you're actually taking an `rf` or `rff` -- the docstring seems to suggest the former while the name suggests the latter

Reply:
right thanks. it's an `rff` the query-processor will call this with the results metadata. and that's the signature of it in fingerprint.clj. I've updated the docstring now.

```clojure
(s/defn ^:private fingerprint-table!
  [table :- i/TableInstance, fields :- [i/FieldInstance]]
  (let [rff (fn [_metadata]
              (redux/post-complete
                (f/fingerprint-fields fields)
                (fn [fingerprints]
                  (reduce (fn [count-info [field fingerprint]]
                            (cond
                              (instance? Throwable fingerprint)
                              (update count-info :failed-fingerprints inc)

                              (some-> fingerprint :global :distinct-count zero?)
                              (update count-info :no-data-fingerprints inc)

                              :else
                              (do
                                (save-fingerprint! field fingerprint)
                                (update count-info :updated-fingerprints inc))))
                          (empty-stats-map (count fingerprints))
                          (map vector fields fingerprints)))))]
    (metadata-queries/table-rows-sample table fields
                                        {:truncation-size truncation-size
                                         :rff             rff})))
```